### PR TITLE
interactive debugger over unix socket over grpc

### PR DESCRIPTION
**Pull request by alexcb**
This changes the interactive debugger to use a unix socket rather than a tcp connection.
The unix socket is proxied over the buildkit session and is encrypted.

This simplifies the debugger as no shellrepeater is required within the
earthly-buildkit container.

This additionally removes the limitation of a single interactive
debugger per host, and allows attaching to remove buildkit instances.

Signed-off-by: Alex Couture-Beil <alex@earthly.dev>

### CODE REVIEW

**Comment by alexcb** [comment]
this requires https://github.com/earthly/buildkit/pull/95

File: debugger/terminal/terminal_other.go [hunk 1]
```diff
@@ -41,20 +41,20 @@ func getWindowSizePayload() ([]byte, error) {
 }
 
 // ConnectTerm presents a terminal to the shell repeater
-func ConnectTerm(ctx context.Context, addr string, console conslogging.ConsoleLogger) error {
-	var d net.Dialer
+func ConnectTerm(ctx context.Context, conn io.ReadWriteCloser, console conslogging.ConsoleLogger) error {
+	//var d net.Dialer
```
Comment (alexcb) [inline debugger/terminal/terminal_other.go:45]: :facepalm: I thought I deleted this code.

File: buildkitd/entrypoint.sh [hunk 2]
```diff
@@ -233,30 +233,21 @@ echo "======== End buildkitd config =========="
...
-        fi
-    fi
+    # shellrepeater is no longer needed -- but perhaps this pattern is useful?
+    #if [ "$shellrepeaterpid" != "0" ]; then
+    #    if ! kill -0 "$shellrepeaterpid" >/dev/null 2>&1; then
+    #        wait "$shellrepeaterpid"
+    #        echo "Error: shellrepeater process has exited with code $?"
+    #        exit 20
+    #    fi
+    #fi
```
Comment (vladaionescu) [inline buildkitd/entrypoint.sh:250]: I would just remove this altogether. We can always pull it back from git history in the future.

File: debugger/terminal/terminal_other.go [hunk 3]
```diff
@@ -41,20 +41,20 @@ func getWindowSizePayload() ([]byte, error) {
...
+	//defer conn.Close()
 
-	_, err = conn.Write([]byte{common.TermID})
-	if err != nil {
-		return errors.Wrap(err, "failed to write TermID connection")
-	}
+	//_, err = conn.Write([]byte{common.TermID})
+	//if err != nil {
+	//	return errors.Wrap(err, "failed to write TermID connection")
+	//}
```
Comment (vladaionescu) [inline debugger/terminal/terminal_other.go:57]: Remove leftover commented code?

File: earthfile2llb/converter.go [hunk 4]
```diff
@@ -1691,6 +1692,19 @@ func (c *Converter) internalRun(ctx context.Context, opts ConvertRunOpts) (pllb.
...
+		if err != nil {
+			_, isCapError := err.(*apicaps.CapError)
+			if !isCapError {
+				c.opt.Console.Warnf("failed to check LLBCaps for CapExecMountSock: %v", err) // keep going
+			} else if opts.Interactive {
+				// only display warning when user supplies --interactive to an older buildkit version
+				c.opt.Console.Warnf("interactive debugger requires a newer version of buildkit: %v", err)
+			}
+		} else {
+			runOpts = append(runOpts, llb.SocketTarget("earthly_interactive", "/var/run/earthly_interactive", 0666, 0, 0))
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:1705]: Presumably, this creates `/var/run/` under the hood if it doesn't exist yet, right?

File: earthfile2llb/converter.go [hunk 5]
```diff
@@ -1691,6 +1692,19 @@ func (c *Converter) internalRun(ctx context.Context, opts ConvertRunOpts) (pllb.
...
 	if !opts.Locally {
 		// Debugger.
+		err := c.opt.LLBCaps.Supports(solverpb.CapExecMountSock)
+		if err != nil {
+			_, isCapError := err.(*apicaps.CapError)
+			if !isCapError {
+				c.opt.Console.Warnf("failed to check LLBCaps for CapExecMountSock: %v", err) // keep going
+			} else if opts.Interactive {
+				// only display warning when user supplies --interactive to an older buildkit version
+				c.opt.Console.Warnf("interactive debugger requires a newer version of buildkit: %v", err)
```
Comment (vladaionescu) [inline earthfile2llb/converter.go:1702]: So this breaks compatibility... but then again nobody is using the debugger in remote buildkit / satellite context today. And for local buildkit, the versions are in lock-step, so it's not an issue.

File: cmd/earthly/build_cmd.go [hunk 6]
```diff
@@ -270,7 +268,7 @@ func (app *earthlyApp) actionBuildImp(cliCtx *cli.Context, flagArgs, nonFlagArgs
 	debuggerSettings := debuggercommon.DebuggerSettings{
 		DebugLevelLogging: app.debug,
 		Enabled:           app.interactiveDebugging,
-		RepeaterAddr:      fmt.Sprintf("%s:8373", bkIP),
+		SocketPath:        "/var/run/earthly_interactive", // TODO const
```
Comment (vladaionescu) [inline cmd/earthly/build_cmd.go:271]: This is a path on the host system, right?

**Review by vladaionescu — COMMENTED** [review]
🎉 

Just some minor remarks/questions...

File: earthfile2llb/converter.go (hunk 5 above)
Comment (alexcb) [reply earthfile2llb/converter.go:1702]: my thinking exactly!

**Review by vladaionescu — APPROVED** [review]
Pre-approving, pending the minor leftover stuff and buildkit PR merge.

File: earthfile2llb/converter.go (hunk 4 above)
Comment (alexcb) [reply earthfile2llb/converter.go:1705]: good question! I took it for granted, but here's a quick manual test:

busybox doesn't have `/var/run`:
```
$ docker run --rm busybox:latest ls -la /var/run
ls: /var/run: No such file or directory
```

but buildkit will create it:

```
$ ./build/linux/amd64/earthly -i ~/test/socket-server+novarrun 
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)

 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

      busybox:latest | --> Load metadata linux/amd64
/h/a/t/socket-server+novarrun | --> FROM busybox:latest
/h/a/t/socket-server+novarrun | [██████████] 100% resolve docker.io/library/busybox:latest@sha256:20142e89dab967c01765b0aea3be4cec3a5957cc330f061e5503ef6168ae6613
/h/a/t/socket-server+novarrun | --> RUN --no-cache false
/h/a/t/socket-server+novarrun | Entering interactive debugger
/h/a/t/socket-server+novarrun | Command /bin/sh -c false failed with exit code 1
/h/a/t/socket-server+novarrun | 

/ # ls -la /var/run
total 8
drwxr-xr-x    2 root     root          4096 Sep 13 00:00 .
drwxr-xr-x    1 root     root          4096 Sep 13 00:00 ..
srw-rw-rw-    1 root     root             0 Sep 13 00:00 earthly_interactive
```

File: cmd/earthly/build_cmd.go (hunk 6 above)
Comment (alexcb) [reply cmd/earthly/build_cmd.go]: it's actually under the container; I forgot about my TODO, hopefully it's more clear now.

**Pull request merged**